### DISCUSSION TITLE
Corne Layer

### DISCUSSION THREAD

**Discussion by Arne-Berner** [dariogoetz/keyboard_layout_optimizer#63] [Q&A] [answered]
Hejhej,
I really like this project, it seems really fun. But I'm not sure if i got it right:
I use a corne keyboard, and I would love to optimize some layout for me using some german and english text. But I am happy about my second and third layer. How would I go about just optimizing my standard layer? I also use keys that are different depending on how long you press them, would that be a problem? (I could just ignore them in that case)
I have little rust experience, if that helps answering the question. If I just skipped some documentation I should have, then I'd love to see that too :)

**Comment by dariogoetz** [comment, marked as answer]
For the corne keyboard, you can start with the `config/keyboard/crkbd.yml` keyboard config file (use the `-l` parameter).

In the `base_layout` section of that file, you find a description of the starting point for the layout. It contains a list of symbols (i.e. characters) for each key of the keyboard (starting with the top-left key going to the left and then row-wise down). The symbols in these lists define what that key produces on the various layers (the first element corresponding to the base layer, when no modifier is pressed, the second element corresponding to the layer that can be activated as described in the first element of the `modifiers` list of the config file.

This means that you should fill all symbols that your layout shall be able to generate into the `base_layout` section. Usually one puts all "alpha" letters (a-z) as first elements and uppercase characters as second elements. Then the symbols of the second and third layers that you are "happy about" should be filled in accordingly.

When you specify a layout to be evaluated (or perform an optimization) and you would like your second and third layer to not "move" with the specified permutation (stay where they are defined in the `base_layout`), then set the corresponding layers in the `fixed_layers` field of the config file (layers start with 0 for the base layer).

The default configuration comes from the neo2 layout (www.neo-layout.org) and leaves the symbol layer (in neo-language called layer 3, here, index 2) and the number/navigation layer (in neo-language called layer 4, here index 3) fixed so that only the alphas, their uppercase variants and the corresponding math-symbols on the upper layers move with the layout permutation.

If you only need alphas, uppercase letters, and your "second and third" layer, the `base_layout` lists shall be of length 4 and if only the alphas should be "optimized" together with their uppercase variants while leaving the "second and third" layer as they are, use `fixed_layers: true`.

I have now added a new modifier type called `long_press` that can generate a different symbol if a key is pressed for a long time. The so-generated symbols shall reside in their own layer. I am not sure, if you already need this for one of the four layers described above; if not, then you can add a fifth item to the `base_layout` lists and have a fourth modifier in `modifiers` with `type: long_press` (and no `value` field).

This is only a first "rough" implementation, though. In particular, you should think about how you would like to define costs that are associated with such a long-press. Currently, they will only be accounted for in the `modifier_usage` metric. There you can define a cost value that is added for each occurrence of a symbol of the long-press layer. This is certainly not a very good metric, yet. In particular, it will always give the same cost no matter the layout permutation. One option would be to take the `key_costs` into account to make long-presses of hard-to-reach keys more expensive than ones in easy locations. This should be discussed.

So far, you should be able to use the program without rust skills. These would be required, if you want to add/change how the individual evaluation metrics shall work. There is only some rust docs that you could generate. As I find less and less time, however, (and had to be time-efficient throughout the whole project), the documentation is missing many newer features and is not maintained very well. Sorry about that.

**Reply by Arne-Berner** [reply]
Thanks a lot for the fast answer and elaborate answer! 👍
